### DISCUSSION TITLE
Errors appearing when running flet on Android when the flet code contains some libraries even though they are binary on https://pypi.org/ like pyzbar:

### DISCUSSION THREAD

**Discussion by YasseenAlhusammi** [flet-dev/flet#4163] [Q&A] [answered]
### Discussed in https://github.com/flet-dev/flet/discussions/4158

<div type='discussions-op-text'>

<sup>Originally posted by **YasseenAlhusammi** October 14, 2024</sup>
Please add pyzbar:
The following reasons:
When converting a Fleet project using to apk using:
Version: flet==0.25.0.dev3519
and python==3.12.6:
and windows 10:
The following error appears on Android when running the application on Android despite adding pyzbar to requirements.txt, and despite the fact that
pyzbar-0.1.9-py2.py3-none-win_amd64.whl has a binary file at https://pypi.org/:
The error that appears when running the application on Android is as follows:
Traceback (most recent call last):
File "<string>", line 43, in <module>
File "<frozen runpy>", line 229, in run_module
File "<frozen runpy>", line 88, in _run_code
File "/data/user/0/com.flet.myapp/files/flet/app/main.py", line 4, in <module> from pyzbar import pyzbar File "/data/user/0/com.flet.myapp/files/flet/python_site_packages/pyzbar/pyzbar.py", line 7, in <module> from .wrapper import ( File "/data/user/0/com.flet.my app/files/flet/python_site_packages/pyzbar/wrapper.py", line 151, in <module> zbar_version = zbar_function( ^^^^^^^^^^^^^^ File "/data/user/0/com.flet.myapp/files/flet/python_site_packages/pyzbar/wrapper.py", line 148, in zbar_function return prototype((fname, load_libzbar())) ^^^^^^^^^^^^^ File "/data/user/0/com.flet.myapp/files/flet/python_site_packages/pyzbar/wrapper.py", line 127, in load_libzbar libzbar, dependencies = zbar_library.load() ^^^^^^^^^ File "/data/user/0/com.flet.myapp/files/flet/python_site_packages/pyzbar/zbar_library.py", line 65, in load raise ImportError('Unable to find zbar shared library') ImportError: Unable to find zbar shared library</div>

**Comment by FeodorFitsner** [comment]
`pyzbar` module depends on `zbar` library written in C. Maybe zbar lib could be compiled for Android, but then probably pyzbar package itself should be modified to be able to load zbar at runtime.
I'm moving that to discussions.

**Comment by YasseenAlhusammi** [comment]
I have a suggestion to modify pyzbar so that it can run zbar:


I found a solution:
Go to https://github.com/NaturalHistoryMuseum/pyzbar/blob/master/pyzbar/zbar_library.py
and edit the file to:

`import platform
import sys
from ctypes import cdll
from ctypes.util import find_library
from pathlib import Path

__all__ = ['load']

def _windows_fnames():
    """Returns the DLL names and dependencies for Windows."""
    if sys.maxsize > 2**32:
        fname = 'libzbar-64.dll'
        dependencies = ['libiconv.dll']
    else:
        fname = 'libzbar-32.dll'
        dependencies = ['libiconv-2.dll']
    return fname, dependencies

def _android_fnames():
    """Returns the library name and dependencies for Android."""
    fname = 'libzbar.so'
    dependencies = ['libiconv.so']  # أضف المكتبة التابعة إذا كانت موجودة
    return fname, dependencies

def load():
    """Loads the libzbar shared library and its dependencies."""
    if platform.system() == 'Windows':
        fname, dependencies = _windows_fnames()

        def load_objects(directory):
            deps = [
                cdll.LoadLibrary(str(directory.joinpath(dep)))
                for dep in dependencies
            ]
            libzbar = cdll.LoadLibrary(str(directory.joinpath(fname)))
            return deps, libzbar

        try:
            dependencies, libzbar = load_objects(Path(''))
        except OSError:
            dependencies, libzbar = load_objects(Path(__file__).parent)
    
    elif platform.system() == 'Linux' and 'ANDROID_ROOT' in os.environ:
        # Assuming we're running on Android
        fname, dependencies = _android_fnames()
        
        def load_objects(directory):
            deps = [
                cdll.LoadLibrary(str(directory.joinpath(dep)))
                for dep in dependencies
            ]
            libzbar = cdll.LoadLibrary(str(directory.joinpath(fname)))
            return deps, libzbar

        try:
            dependencies, libzbar = load_objects(Path(''))
        except OSError:
            dependencies, libzbar = load_objects(Path(__file__).parent)
    
    else:
        path = find_library('zbar')
        if not path:
            raise ImportError('Unable to find zbar shared library')
        libzbar = cdll.LoadLibrary(path)
        dependencies = []

    return libzbar, dependencies`
Including both: libiconv.so and libzbarjni.so after downloading them from the link: [https://github.com/chentao0707/ZBarAndroidSDK/tree/master/ZBarScanProjAll/libs/arm64-v8a](url)
Renaming libZBarDecoder.so to libzbarjni.so
The proposed solution is effective as stated: [https://stackoverflow.com/questions/32426861/zbar-android-arm-64-couldnt-find-libiconv-so](url)
But personally I can't rebuild due to problems with my operating environment, please try the suggestion I think it works, many have praised the suggestion. I tried to implement it:

**Comment by YasseenAlhusammi** [comment]
I've been stuck trying to get pyzbar to work on Android for a whole month now and I still can't. I hope packages are prioritized before anything else so that flet stands out from other tools.

**Comment by YasseenAlhusammi** [comment]
In short: The solution is available at the following link:
https://stackoverflow.com/questions/32426861/zbar-android-arm-64-couldnt-find-libiconv-so
But, I had difficulty implementing the solution, because I am a beginner in programming.

**Comment by usermicrodevices** [comment]
https://pub.dev/packages/flutter_zxing
https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/python/README.md
perhups this is alternative

**Reply by YasseenAlhusammi** [reply]
Thank you for your interest and cooperation. Frankly, the matter has become very frustrating for me. For a whole year, I have not been able to do anything until now. I still check the page every day hoping to find a solution. Thank you from the bottom of my heart for your interest. I will try to implement the solution and will inform you of the results.

**Comment by usermicrodevices** [comment]
for pure python barcode encoding see https://github.com/xulihang/EAN13_Reader. 
now i do small modification and put in my repo https://github.com/usermicrodevices/prod-flet/

**Comment by ndonkoHenri** [comment, marked as answer]
pyzbar has been added for Android and iOS: https://pypi.flet.dev/pyzbar

Give it a try and let know.
